### PR TITLE
Basic functionalities 

### PR DESCRIPTION
**Pull request by tomerwald**
+Connection encryption
+basic command running
+file upload

### CODE REVIEW

File: peer_protocol/message.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package peer_protocol
+
+import "encoding/binary"
+
+type Message struct {
+	Type    byte
+	Payload []byte
+}
+
+func (m Message) Buffer() []byte {
```
Comment (ShahafLevin) [inline peer_protocol/message.go:11]: Add short doc, it might help to understand this function purpose

File: peer_protocol/peer.go [hunk 2]
```diff
@@ -23,59 +59,136 @@ func (p PeerConnection) createMessageBuffer(mt byte, payload []byte) []byte {
...
 	}
-	fmt.Println(l)
 }
 
-func (p PeerConnection) IsHandshakeValid(protocol string, hash []byte) bool {
+func (p *PeerConnection) IsHandshakeValid(protocol string, hash []byte) bool {
 	if protocol != protocol {
 		return false
-	} else if bytes.Compare(hash, p.InfoHash) == 0 {
+	} else if bytes.Compare(hash, p.Config.InfoHash) == 0 {
```
Comment (ShahafLevin) [inline peer_protocol/peer.go]: ``return bytes.Compare(hash, p.Config.InfoHash) != 0``

File: peer_protocol/peer.go [hunk 3]
```diff
@@ -23,59 +59,136 @@ func (p PeerConnection) createMessageBuffer(mt byte, payload []byte) []byte {
...
 		return true
 	}
 }
 
-func (p PeerConnection) ReceiveHandshake() bool {
+func (p *PeerConnection) ReceiveHandshake() bool {
 	fmt.Println("Handshaking")
-	buf := make([]byte, 68)
-	_, err := p.Conn.Read(buf)
+	buf, err := p.readSocket(68, 10)
```
Comment (ShahafLevin) [inline peer_protocol/peer.go]: Seems like you have a lot of magic numbers, it might be hard to follow them this way.

File: peer_protocol/peer.go [hunk 4]
```diff
@@ -23,59 +59,136 @@ func (p PeerConnection) createMessageBuffer(mt byte, payload []byte) []byte {
...
+	messageLength := binary.BigEndian.Uint32(lengthBuffer)
+	if err == nil {
+		return messageLength, nil
+	} else {
+		return 0, err
+	}
+}
+
+func (p *PeerConnection) ReceiveMessage() (Message, error) {
+	messageLength, lerr := p.ReceiveLength()
```
Comment (ShahafLevin) [inline peer_protocol/peer.go:117]: why lerr?

File: peer_protocol/piece.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package peer_protocol
```
Comment (ShahafLevin) [inline peer_protocol/piece.go:1]: look at https://golang.org/doc/effective_go.html#package-names for a good example for package names

File: razor/razor.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+	default:
+		return commands.NewCommandError("Unknown command type: " + string(com.Type))
+	}
+	return err
+}
+func (r *Client) MessageCycle() (peer_protocol.Message, error) {
+	if m, err := r.Peer.ReceiveMessage(); err == nil {
+		switch m.Type {
+		case peer_protocol.Bitfield:
+			r.HandleBitField(m)
```
Comment (ShahafLevin) [inline razor/razor.go]: Notice that sometimes the H is upper and sometimes lower. Be consistent

File: razor/razor.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+		case peer_protocol.Choke:
+			r.handleChoke()
+		case peer_protocol.Unchoke:
+			r.handleUnChoke()
```
Comment (ShahafLevin) [inline razor/razor.go:159]: Why not handleUnchoke (as on peer_protocol)?

File: readme.md [hunk 8]
```diff
@@ -1 +1,27 @@
...
+    1. Create an info hash (any sha1 will do)
+    2. Decide on a remote port to listen on
+    3. Create a 256 bit long encryption key
+    4. Compile as exe
+    5. Run PyRazor using your key and hash
+    6. Connect and have fun.
+    
+Minor disclosures:
+* The encryption is done poorly, it's only there so that
+the content of commands and uploaded files will not be in clear text
```
Comment (ShahafLevin) [inline readme.md:22]: plain text :)

File: readme.md [hunk 9]
```diff
@@ -1 +1,27 @@
...
+
+* To a BitTorrent expert the communication will look odd
+but anyone who didn't read the protocol documentation would not find a difference
+     
+
```
Comment (ShahafLevin) [inline readme.md:27]: You can also add a doc for your implementation of the protocol (and a reference to a doc for the protocol)

File: peer_protocol/peer.go [hunk 10]
```diff
@@ -1,20 +1,56 @@
...
+		return buf, err
+	} else {
+		return buf, err
+	}
+}
+
+func (p *PeerConnection) Disconnect() {
+	err := p.Conn.Close()
+	if err != nil {
+		fmt.Printf("failed closing connection with client %s", p.Conn.RemoteAddr().String())
```
Comment (ShahafLevin) [inline peer_protocol/peer.go]: Consider to use log instead of fmt, I think it might be helpful in case you'll later want to write your logs to a file

**Review by ShahafLevin — COMMENTED** [review]
Looks awesome!

File: peer_protocol/peer.go (hunk 10 above)
Comment (tomerwald) [reply peer_protocol/peer.go]: will do
